### PR TITLE
Synchronize distributed generation and execute chat tools on rank zero

### LINKED ISSUE #270: Distributed Torch and Triton sampling can diverge across model-parallel ranks
## Summary

The reference Torch and Triton token generators perform stochastic next-token sampling independently on every distributed rank.

Both model-parallel implementations cooperate across ranks during the MLP and therefore require every rank to continue with the same token sequence. With `temperature > 0`, each rank currently calls `torch.multinomial()` locally and no sampled token is broadcast.

`gpt_oss.chat` invokes `generator.generate(...)` without overriding temperature, so it uses the generators' default `temperature=1.0`. A multi-GPU chat run can therefore choose different next tokens on different ranks after the first sampling step.

## Current behavior

Both generators contain the equivalent of:

```python
probs = torch.softmax(logits * (1.0 / temperature), dim=-1)
predicted_token = torch.multinomial(probs, num_samples=1).item()
```

There is no distributed synchronization of `predicted_token` before it becomes the next input token.

## Impact

If model-parallel ranks sample different tokens, subsequent forward passes combine shard computations for different token histories. The distributed inference result is no longer a coherent model sequence and can become nondeterministic/corrupted under the default interactive-chat sampling path.

Greedy `temperature=0` generation happens to avoid the divergence because identical logits produce the same argmax on each rank.

## Proposed resolution

When `torch.distributed` is initialized, sample/select the next token on rank 0 and broadcast that token to every rank before continuing generation. Keep the existing local sampling path when distributed inference is not initialized.

Apply the same synchronization policy to both the Torch and Triton reference generators and add CPU-only mocked regressions showing nonzero ranks use the token supplied by rank 0 rather than sampling independently.

### LINKED ISSUE #290: Distributed chat executes each tool call independently on every model-parallel rank
## Summary

`gpt_oss.chat` broadcasts user input across distributed ranks, but tool/function execution is not rank-gated. When Torch or Triton chat is launched with `torchrun`, every rank enters the same browser, Python, or `apply_patch` branch and executes the tool independently.

## Impact

A single model tool call can therefore produce N external side effects on N model-parallel ranks:

- browser searches/API calls are duplicated;
- Python tool work is duplicated;
- `apply_patch` can race multiple processes against the same working tree;
- non-deterministic tool results can leave ranks with different conversation histories before the next model-parallel generation step.

`get_user_input()` already establishes the intended rank-0/broadcast pattern for interactive input; tool results need the same treatment.

## Proposed resolution

Execute each tool/function call only on rank 0, serialize Harmony `Message` results with `Message.to_dict()`, broadcast the result payload, and reconstruct it with `Message.from_dict()` on every rank. Propagate rank-0 execution failures to every rank as well so nonzero ranks cannot block forever waiting for a result.

Add mocked distributed regression coverage proving nonzero ranks do not execute the tool callback and receive the rank-0 result.

### PR DESCRIPTION
**Pull request by sylvesterkaczmarek**
## Summary

Keep model-parallel chat ranks on one token history and execute external tool/function calls only once.

Two distributed paths were independently unsafe:

- Torch and Triton ranks sampled stochastic next tokens independently, so generated histories and assistant actions could diverge;
- browser, Python, and `apply_patch` calls executed independently on every rank, duplicating external side effects.

Fixes #270.
Fixes #290.

## Fix

- sample/select each next token on rank 0 and broadcast it to all ranks in both Torch and Triton reference generators;
- in single-process mode, preserve existing greedy and stochastic sampling behavior;
- execute browser, Python, and `apply_patch` callbacks only on rank 0;
- serialize Harmony tool-result `Message` objects with `to_dict()`, broadcast them, and reconstruct them with `Message.from_dict()` on every rank;
- broadcast rank-0 tool execution failures so peer ranks fail together rather than blocking.

## Regression coverage

Adds CPU-only mocked tests verifying that:

- local greedy sampling is unchanged;
- rank 0 samples and broadcasts its token;
- nonzero ranks never sample independently;
- local tool execution is unchanged;
- nonzero ranks never invoke tool callbacks and receive rank-0 messages;
- rank-0 tool failures propagate to peers.

Model logits, temperature scaling, stop-token handling, tool schemas, and non-distributed chat behavior are otherwise unchanged.

### CODE REVIEW

File: gpt_oss/chat.py [hunk 1]
```diff
@@ -58,6 +58,35 @@ def get_user_input():
...
+                "error": None,
+            }
+        except Exception as exc:
+            payload = {
+                "messages": [],
+                "error": f"{type(exc).__name__}: {exc}",
+            }
+
+    payload_list = [payload]
+    torch.distributed.broadcast_object_list(payload_list, 0)
```
[the remark this answers is not in the document]
Comment (sylvesterkaczmarek) [reply gpt_oss/chat.py:80]: Good catch. Fixed in 397fee3: the PR now also synchronizes Torch and Triton next-token sampling by selecting the token on rank 0 and broadcasting it before generation continues. This removes the pre-tool action divergence, and the tool-result broadcast now runs only after every rank is on the same token/action history. The sampling regression coverage from #285 is included here as well.

**Comment by sylvesterkaczmarek** [comment]
@codex review
